Initial support for Wininet.dll and enumeration of its' / IE's cache

mlfreeman2 opened the pull request:
Adds Wininet interface with the following functions in Wininet.dll:
* `FindFirstUrlCacheEntry`
* `DeleteUrlCacheEntry` 
* `FindCloseUrlCache`
* `FindNextUrlCacheEntry`

Also includes the `INTERNET_CACHE_ENTRY_INFO` structure in Wininet.h

Also includes a helper function for parsing WinInet's / IE's cache.

And last but not least, also includes tests for helper util method and all individual functions in Wininet.

dblock commented on contrib/platform/test/com/sun/jna/platform/win32/WininetUtilTest.java:
```diff
@@ -0,0 +1,71 @@
+                googleLogoOrOtherImageFound);
+        assertTrue("History entry for google.com should have been found in the browser cache.", historyEntryFound);
+
+        Runtime.getRuntime().exec("taskkill.exe /f /im iexplore.exe");
```
Isn't this going to kill anyone's internet explorer who's writing JNA code in development and running this test? Probably not a great idea?

dblock commented on contrib/platform/test/com/sun/jna/platform/win32/WininetUtilTest.java:
```diff
@@ -0,0 +1,71 @@
+        Map<String, String> ieCache = WininetUtil.getCache();
+        assertTrue("WinInet cache should have some items in it.", ieCache.size() > 1);
+
+        boolean historyEntryFound = false;
```
It's a nice test, but I wonder if we're trying too hard :)

dblock commented:
Nice work @mlfreeman2, just see my question on killing IE processes.

lgoldstein commented on contrib/platform/src/com/sun/jna/platform/win32/Wininet.java:
```diff
@@ -0,0 +1,372 @@
+     * is exempt from scavenging.
+     */
+    public static final int EDITED_CACHE_ENTRY        = 8;
+
```
No need for the _public static final_ please remove them.

lgoldstein commented on contrib/platform/src/com/sun/jna/platform/win32/Wininet.java:
```diff
@@ -0,0 +1,372 @@
+     *         </li>
+     *         </ul>
+     */
+    public boolean DeleteUrlCacheEntry(String lpszUrlName);
```
No need for _public_ - all methods in an interface are such - please remove the modifier in all the methods

dblock replied:
:+1:

dblock commented:
@lgoldstein is right wrt public and such, I didn't pay attention. Thanks.

lgoldstein commented on contrib/platform/src/com/sun/jna/platform/win32/Wininet.java:
```diff
@@ -0,0 +1,372 @@
+                    "CacheEntryType", "dwUseCount", "dwHitRate", "dwSizeLow", "dwSizeHigh", "LastModifiedTime",
+                    "ExpireTime", "LastAccessTime", "LastSyncTime", "lpHeaderInfo", "dwHeaderInfoSize",
+                    "lpszFileExtension", "u", "additional" });
+        }
```
Don't suppress the warning and no need to allocate _new String[]_ - use
```java
@Override
protected List<String> getFieldOrder() {
    return Arrays.asList("field1", "field2", ...);
}
```

mlfreeman2 commented:
@dblock - I addressed the feedback from @lgoldstein and put constraints on what IE windows get killed off now. 

On my box, I opened non-Google IE windows and then ran the test - and the test opened Google and killed it while leaving non-Google IE windows alone. 

In theory someone who had an existing google.com window open could be trouble - if that's a concern I'll have to get more creative.

mlfreeman2 commented:
@lgoldstein - I saw that your PR for generics with `Native.loadLibrary()` was accepted, so I updated my `INSTANCE` object in `com.sun.jna.platform.win32.Wininet` accordingly - if there are other adjustments let me know.

dblock commented:
This looks good to me, @mlfreeman2 it needs to be reabased, @lgoldstein feel free to merge.
